File: day6/day6.js
const fs = require('fs');
const input = fs.readFileSync("day6/day6.txt", 'utf-8').split('\n');

function part1() {
    var grid = Array.from({length: 1000}, () => (Array.from({length: 1000}, () => (0))));
    input.forEach(function(line) {
        var items = line.split(" ");
        if (items[0] == "turn") items.splice(0, 2, items[0] + " " + items[1]);
        items.splice(2, 1);
        items[1] = items[1].split(",").map(s => +s);
        items[2] = items[2].split(",").map(s => +s);
        switch(items[0]) {
            case "toggle":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] = grid[i][j] == 0 ? 1 : 0;
                break;
            case "turn on":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] = 1;
                break;
            case "turn off":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] = 0;
                break;
        }
    });
    return grid.map(l => l.reduce((a, b) => a + b, 0)).reduce((a, b) => a + b, 0);
}

function part2() {
    var grid = Array.from({length: 1000}, () => (Array.from({length: 1000}, () => (0))));
    input.forEach(function(line) {
        var items = line.split(" ");
        if (items[0] == "turn") items.splice(0, 2, items[0] + " " + items[1]);
        items.splice(2, 1);
        items[1] = items[1].split(",").map(s => +s);
        items[2] = items[2].split(",").map(s => +s);
        switch(items[0]) {
            case "toggle":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] += 2;
                break;
            case "turn on":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] += 1;
                break;
            case "turn off":
                for (var i = items[1][1]; i <= items[2][1]; i++)
                    for (var j = items[1][0]; j <= items[2][0]; j++)
                        grid[i][j] = Math.max(0, grid[i][j] - 1);
                break;
        }
    });
    return grid.map(l => l.reduce((a, b) => a + b, 0)).reduce((a, b) => a + b, 0);
}

console.log("Part 1: " + part1());
console.log("Part 2: " + part2());

File: day6/day6.py
data = open('day6/day6.txt').read().split('\n')

def part1():
    grid = [[0] * 1000 for _ in range(1000)]
    for line in data:
        items = line.split(' ')
        if items[0] == 'turn':
            items[0:2] = [' '.join(items[0:2])]
        items.pop(2)
        for i in range(1, 3):
            items[i] = [int(n) for n in items[i].split(',')]
        if items[0] == 'toggle':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] = int(not bool(grid[i][j]))
        elif items[0] == 'turn on':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] = 1
        elif items[0] == 'turn off':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] = 0
    return sum(map(lambda g: sum(g), grid))

def part2():
    grid = [[0] * 1000 for _ in range(1000)]
    for line in data:
        items = line.split(' ')
        if items[0] == 'turn':
            items[0:2] = [' '.join(items[0:2])]
        items.pop(2)
        for i in range(1, 3):
            items[i] = [int(n) for n in items[i].split(',')]
        if items[0] == 'toggle':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] += 2
        elif items[0] == 'turn on':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] += 1
        elif items[0] == 'turn off':
            for i in range(items[1][1], items[2][1] + 1):
                for j in range(items[1][0], items[2][0] + 1):
                    grid[i][j] = max(0, grid[i][j] - 1)
    return sum(map(lambda g: sum(g), grid))

print(f'Part 1: {part1()}')
print(f'Part 2: {part2()}')